packages/runtime/container-runtime/src/summary/summaryGenerator.ts

```diff
@@ -407,78 +407,94 @@ export class SummaryGenerator {
+			const summaryDuration = Date.now() - this.heuristicData.lastAttempt.summaryTime;
 
-			// adding new properties
-			summarizeTelemetryProps = {
-				ackWaitDuration: ackNackDuration,
-				ackNackSequenceNumber: ackNackOp.sequenceNumber,
-				summarySequenceNumber: ackNackOp.contents.summaryProposal.summarySequenceNumber,
-				...summarizeTelemetryProps,
-			};
-			if (ackNackOp.type === MessageType.SummaryAck) {
+			// Declaring a re-usable function here which can be called either after a summary boradcast is received or after summaryAck is received
+			// based on whether client is uses a single-commit summary pattern or two-commit summary respectively.
+			const afterSuccessfullSummary = async (opReceivedForSuccessfullSummary) => {
 				this.heuristicData.markLastAttemptAsSuccessful();
 				this.successfulSummaryCallback();
 				summarizeEvent.end({
 					...summarizeTelemetryProps,
-					handle: ackNackOp.contents.handle,
+					handle: opReceivedForSuccessfullSummary.contents.handle,
 				});
 				// This processes the summary ack of the successful summary. This is so that the next summary does not
 				// start before the ack of the previous summary is processed.
 				await this.refreshLatestSummaryCallback({
 					proposalHandle: summarizeOp.contents.handle,
-					ackHandle: ackNackOp.contents.handle,
+					ackHandle: opReceivedForSuccessfullSummary.contents.handle,
 					summaryRefSeq: summarizeOp.referenceSequenceNumber,
 					summaryLogger,
 				});
 				resultsBuilder.receivedSummaryAckOrNack.resolve({
 					success: true,
 					data: {
-						summaryAckOp: ackNackOp,
-						ackNackDuration,
+						summaryAckOp: opReceivedForSuccessfullSummary,
+						ackNackDuration: summaryDuration,
 					},
 				});
-			} else {
-				// Check for retryDelay in summaryNack response.
-				assert(ackNackOp.type === MessageType.SummaryNack, 0x274 /* "type check" */);
-				const summaryNack = ackNackOp.contents;
-				const errorMessage = summaryNack?.message;
-				const retryAfterSeconds = summaryNack?.retryAfter;
-
-				// pre-0.58 error message prefix: summaryNack
-				const error = new LoggingError(`Received summaryNack`, {
-					retryAfterSeconds,
-					errorMessage,
-				});
+			};
 
-				assert(
-					getRetryDelaySecondsFromError(error) === retryAfterSeconds,
-					0x25f /* "retryAfterSeconds" */,
-				);
-				// This will only set resultsBuilder.receivedSummaryAckOrNack, as other promises are already set.
-				return fail(
-					"summaryNack",
-					error,
-					{ ...summarizeTelemetryProps, nackRetryAfter: retryAfterSeconds },
-					undefined /* submitFailureResult */,
-					{ summaryNackOp: ackNackOp, ackNackDuration, retryAfterSeconds },
+			// includesProtocolTree property represents the summary was generated using single-commit summary pattern
+			// thus there is no need to wait for ack/nacks and we can simply use the summarizeOp to extract the summary handle.
+			if (summarizeOp.contents.details?.includesProtocolTree) {
```

Comment:
I do not remember all the details by now, but I'd double check if that property signifies single-commit summary, or just client telling service that it's Ok with single-commit summary. In the latter case you likely can't assume that we can skip summaryAck.
This is mostly related to the order we make changes in the system. If we wait for service to completely deploy all of their changes first, and client controls what kind of summary is generated (by this flag), then it's safe to pivot based on that. But if client starts uploading full summaries, but service either does not have bits to produce single summary, or rollout / killbit switches are controlled from service, then we need some return signal from service here.

Note that it will be harder with AFR. With ODSP + Loop, we controlled full flow. So, we could make a call that deployment is controlled from client (including killbit feature gates to disable if something goes wrong). We will likely need to control everything from service side for AFR, as we cannot control feature gates for 3rd party clients, and we can't be sure system works fine when we release client bits - we do not have enough internal traffic to be sure everything is fine.

This brings two key follow ups IMHO (and we touched on that last Wed in one of the meetings with Navin) - one of these should be true:
1. We need to make sure that 2.0 is single-commit ready. I.e. it will use single commit everywhere where service allows us to do so, including in AFR.
2. We have an agreement with PowerPlatform (biggest AFR customer) to help us in delivering single-commit for AFR. I.e. we could given them a client config (and they control it through their feature gates) to enable single-commit somewhere in 2.x timeframe (i.e. we control rollout from client, and PowerPlatform is the way how we test-flight it, before making it ON for all customers in 3.0 timeframe).

Reply:
I checked with Gary and he confirmed that `includesProtocolTree` property signifies single-commit summary was used. Infact it is also the property that ODSP service uses to reject the two-commit summary cases. (@GaryWilber  to keep me honest here). As of January this year, all the two-commit summaries are being rejected and all odsp-clients are on single-commit summaries.

@vladsud With this information, do you still think this solution will be shaky?

Reply:
I do not think it's shaky, we just do not have a plan how we will get there for AFR, that's it.

It might be fine as is, but we likely need a new mechanism for client to know if service supports single-commit summaries. In other words, instead of driving single-commit rollout from client (how we did it for ODSP, using "Fluid.Container.summarizeProtocolTree2" feature gate and ILoaderOptions.summarizeProtocolTree), we will likely drive it from service. And thus, we need some mechanism for client to know if it can upload full summary, and if it can set this flag on summary op.
With ODSP it was rather simple - all service code was deployed first, so client could rely on that. With AFR, we likely need to assume that service can turn on/off single-commit, and whole system should adjust based on that.

Maybe other way to say it - while we are in transition (i.e. there are both single-commit and two-commit summaries possible), includesProtocolTree could mean - single summary process was used from the client (i.e. full snapshot was uploaded to storage), but it might still result in two-commit process because service decided so.
With this change we remove that possibility.
But it's neither bad nor good - we do not know (until we build up deployment plan) whether we rely on such state / possibility.